### PR TITLE
[Bugfix][KV Offload] Fall back when MADV_POPULATE_WRITE is unsupported

### LINKED ISSUE #44888: [Bug]: vllm start with TieringOffloadingSpec mmap_obj error
### Your current environment

vllm v0.22.0

--kv-transfer-config '{"kv_connector":"OffloadingConnector","kv_role":"kv_both","kv_connector_extra_config":{"spec_name":"TieringOffloadingSpec","cpu_bytes_to_use":3221225472,"secondary_tiers":[{"type":"fs_python","root_dir": "/data01/offload"}]}}' 

### 🐛 Describe the bug

```
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) INFO 06-08 14:19:55 [v1/kv_offload/cpu/shared_offload_region.py:68] Created mmap file /dev/shm/vllm_offload_1780928262818256525_dp0.mmap (3.22 GB)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962] WorkerProc hit an exception.
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962] Traceback (most recent call last):
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/executor/multiproc_executor.py", line 957, in worker_busy_loop
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     output = func(*args, **kwargs)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]              ^^^^^^^^^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/worker_base.py", line 319, in initialize_from_config
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self.worker.initialize_from_config(kv_cache_config)  # type: ignore
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/tracing/otel.py", line 178, in sync_wrapper
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     return func(*args, **kwargs)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]            ^^^^^^^^^^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_worker.py", line 558, in initialize_from_config
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self.model_runner.initialize_kv_cache(kv_cache_config)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_model_runner.py", line 7191, in initialize_kv_cache
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     kv_transfer_group.register_kv_caches(kv_caches)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py", line 76, in register_kv_caches
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self.connector_worker.register_kv_caches(kv_caches)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py", line 250, in register_kv_caches
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self._register_handlers(canonical_kv_caches)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py", line 56, in _register_handlers
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     for src_cls, dst_cls, handler in self.spec.get_handlers(kv_caches):
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/kv_offload/cpu/spec.py", line 99, in get_handlers
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self._handlers = self.create_handlers(kv_caches)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]                      ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/kv_offload/tiering/spec.py", line 171, in create_handlers
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     worker_mmap = SharedOffloadRegion(
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]                   ^^^^^^^^^^^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/kv_offload/cpu/shared_offload_region.py", line 47, in __init__
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self.mmap_obj.madvise(
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     ~~~~~~~~~~~~~~~~~~~~^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962] OSError: [Errno 22] Invalid argument
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962] Traceback (most recent call last):
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/executor/multiproc_executor.py", line 957, in worker_busy_loop
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     output = func(*args, **kwargs)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]              ^^^^^^^^^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/worker_base.py", line 319, in initialize_from_config
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self.worker.initialize_from_config(kv_cache_config)  # type: ignore
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/tracing/otel.py", line 178
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     return func(*args, **kwargs)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]            ^^^^^^^^^^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_worker.py",
line 558, in initialize_from_config
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self.model_runner.initialize_kv_cache(kv_cache_config)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/worker/gpu_model_runner
.py", line 7191, in initialize_kv_cache
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     kv_transfer_group.register_kv_caches(kv_caches)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/distributed/kv_transfer/kv
_connector/v1/offloading_connector.py", line 76, in register_kv_caches
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self.connector_worker.register_kv_caches(kv_caches)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/distributed/kv_transfer/kv
_connector/v1/offloading/worker.py", line 250, in register_kv_caches
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self._register_handlers(canonical_kv_caches)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/distributed/kv_transfer/kv
_connector/v1/offloading/worker.py", line 56, in _register_handlers
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     for src_cls, dst_cls, handler in self.spec.get_handlers(kv_caches):
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/kv_offload/cpu/spec.py"
, line 99, in get_handlers
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]     self._handlers = self.create_handlers(kv_caches)
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]                        ^^^^^^^^^^^^^^^^^^^^^^^^^^
(API Server pid=11) (RayWorkerProc pid=12505) (Worker_DPO_TP6_EP6 pid=12505) ERROR 06-08 14:19:55 [v1/executor/multiproc_executor.py:962]   File "/usr/local/lib/python3.12/dist-packages/vllm/v1/kv_offload/tiering/spec
.py", line 171, in create_handlers
(API Server pid=11) (DPMoEEEngineCoreActor pid=319, ) Exception raised in creation task: The actor died because of an error raised in its creation task, ray::DPMoEEEngineCoreActor.__init__() (pid=319, ip
=10.127.174.15, actor_id=4165e141e7e2751972a8691201000000, repr=vllm.v1.engine.core.DPMoEEEngineCoreActor object at 0x7f872e51e8d0>)

  File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 2155, in __init__
    DPEngineCoreProc.__init__(
  File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 1689, in __init__
    super().__init__(
  File "/usr/local/lib/python3.12/dist-packages/vllm/tracing/otel.py", line 178, in sync_wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 905, in __init__
    super().__init__(
  File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 131, in __init__
    kv_cache_config = self._initialize_kv_caches(vllm_config)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/vllm/tracing/otel.py", line 178, in sync_wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/vllm/v1/engine/core.py", line 286, in _initialize_kv_caches
    self.model_executor.initialize_from_config(kv_cache_configs)
  File "/usr/local/lib/python3.12/dist-packages/vllm/v1/executor/abstract.py", line 123, in initialize_from_config
    self.collective_rpc("initialize_from_config", args=(kv_cache_configs,))
  File "/usr/local/lib/python3.12/dist-packages/vllm/v1/executor/multiproc_executor.py", line 403, in collective_rpc
    return future if non_block else future.result()
                                    ^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/vllm/v1/executor/multiproc_executor.py", line 90, in result
    return super().result()
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.12/dist-packages/vllm/v1/executor/multiproc_executor.py", line 94, in _wait_for_response
    response = self.aggregate(self.get_responses())
  File "/usr/local/lib/python3.12/dist-packages/vllm/v1/executor/multiproc_executor.py", line 390, in get_response
    raise RuntimeError(
RuntimeError: Worker failed with error '[Errno 22] Invalid argument', please check the stack trace above for the root cause
```
Image

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by Alex-ai-future**
Fixes #44888

## Why this PR

`SharedOffloadRegion.__init__` unconditionally issues `madvise(MADV_POPULATE_WRITE, …)` on its shared mmap. On Linux < 5.14 (RHEL 8 ships 4.18) the syscall returns `EINVAL`, so any vLLM startup that touches `TieringOffloadingSpec` aborts before KV offloading can register.

This PR probes the advice once after mmap creation and falls back to a no-mutation NumPy read-modify-write (`|= 0`) that triggers the same page fault without altering any existing bytes. Other `OSError`s (e.g. `EIO`) propagate.

## Changes

`vllm/v1/kv_offload/cpu/shared_offload_region.py` (+32 / -6):

- Module-level helpers `_madvise_populate_write` / `_fallback_populate_write` are the test seams (`mmap.mmap` is an immutable C type on Python 3.12+ and cannot be subclassed/patched).
- `_get_populate_write_fn(mmap_obj)` probes once via `_madvise_populate_write(mmap_obj, 0, PAGESIZE)`. On `OSError(EINVAL)`, logs once and returns `_fallback_populate_write` (which uses `arr[::PAGESIZE] |= 0` — no byte mutation, still faults in the page). Other `OSError`s propagate.
- The ranked and unranked call sites in `__init__` now invoke the returned helper instead of `mmap_obj.madvise(...)` directly. Behaviour on kernels ≥ 5.14 is unchanged.

`tests/v1/kv_offload/cpu/test_shared_offload_region.py` (+107 / -6):

- `test_madvise_success_selects_madvise_population` — spies on both helpers; asserts 1 probe + N populate calls, fallback 0 times.
- `test_madvise_einval_smoke` — constructor completes when the probe raises `EINVAL`.
- `test_fallback_populate_write_preserves_existing_bytes` — plants non-zero sentinels at each page head, drives `_fallback_populate_write` directly, asserts the sentinels survive. This is the test that distinguishes `= 0` from `|= 0`.
- `test_madvise_unexpected_oserror_propagates` — non-`EINVAL` `OSError`s propagate out of `__init__`.

`#44913` (guygir, OPEN since 2026-06) targets the same bug but was stalled on pre-commit / CI failures caused by `mmap.mmap` subclassing. The three commits on top of its merged base (`5031c2e`) — `990b8d6` (Python 3.12+ test rewrite), `10f7bfb` (`|= 0` per @orozery's review), `3233a6f` (route the probe through `_madvise_populate_write` so the module-level test seam actually intercepts it) — address those root causes.

## Test results

The submitter ran `tests/v1/kv_offload/cpu/test_shared_offload_region.py -v` on two kernel versions:

- **Linux 5.14+** (development host, kernel accepts `MADV_POPULATE_WRITE`): 32 passed in 16.11s. Confirms the probe succeeds and the native path is selected.
- **Linux 5.4.0-155-generic** (Ubuntu, kernel rejects `MADV_POPULATE_WRITE` with `EINVAL`): 32 passed in 26.94s. Confirms end-to-end that the probe routes through `_madvise_populate_write`, the kernel returns `EINVAL`, and the `|= 0` fallback path executes without altering existing bytes.

```bash
pre-commit run --files \
    vllm/v1/kv_offload/cpu/shared_offload_region.py \
    tests/v1/kv_offload/cpu/test_shared_offload_region.py
# 12 hooks PASSED
```

No model-evaluation results are reported because this change does not affect model output, accuracy, or serving throughput — it only adds a fallback for the mmap pre-fault on kernels that lack the advice.

## AI assistance

This PR was prepared with the assistance of Claude (Anthropic). All code changes were reviewed line-by-line by the human submitter and the relevant tests were run on a Linux runner before submission, per the project's contribution policy.

### CODE REVIEW

**Comment by Alex-ai-future** [comment]
@orozery  PTAL

**Comment by orozery** [comment]
@Etelis PTAL
I think that #44913 approach is cleaner but it needs fixing.
Maybe we want to poke Guy?

**Comment by Alex-ai-future** [comment]
> @Etelis PTAL I think that #44913 approach is cleaner but it needs fixing. Maybe we want to poke Guy?

@orozery Thanks
should I close this PR in favor of https://github.com/vllm-project/vllm/pull/44913?

**Comment by Etelis** [comment]
I agree. @orozery 

@Alex-ai-future do you maybe want to hop in and assist there?

**Comment by Alex-ai-future** [comment]
> I agree. @orozery
> 
> @Alex-ai-future do you maybe want to hop in and assist there?

Sure, happy to. Will fetch @guygir's branch

**Comment by Alex-ai-future** [comment]
Done — force-pushed the branch with guygir's commits preserved (merge of guygir/vllm#44913) plus one commit on top to make the fallback tests Py 3.12+ compatible. Root cause was `mmap.mmap` being an immutable built-in on 3.12+, so subclassing / patching the type fails with TypeError before any test body runs — which is why the mergify bot kept flagging pre-commit. Tests now monkeypatch the module-level helpers instead.

**Comment by Alex-ai-future** [comment]
pre-commit failed for other problems

File: vllm/v1/kv_offload/cpu/shared_offload_region.py [hunk 1]
```diff
@@ -25,6 +31,32 @@ def _wait_for_file_size(fd: int, expected_size: int, timeout: float = 30.0) -> N
         time.sleep(0.005)
 
 
+def _madvise_populate_write(mmap_obj: mmap.mmap, offset: int, length: int) -> None:
+    mmap_obj.madvise(_MADV_POPULATE_WRITE, offset, length)
+
+
+def _fallback_populate_write(mmap_obj: mmap.mmap, offset: int, length: int) -> None:
+    arr = np.frombuffer(mmap_obj, dtype=np.uint8)
+    arr[offset : offset + length : mmap.PAGESIZE] = 0
```
Comment (orozery) [inline vllm/v1/kv_offload/cpu/shared_offload_region.py]: Let's restore to your proposal of ` |= 0` ?

File: tests/v1/kv_offload/cpu/test_shared_offload_region.py [hunk 2]
```diff
@@ -404,6 +405,94 @@ def test_file_has_correct_size(iid):
...
+        for page in (1, 3, 5):
+            assert raw[page * PAGE_SIZE] == 0, (
+                f"fallback must touch rank-1 page {page} head"
+            )
+            assert all(
+                b == 0 for b in raw[page * PAGE_SIZE + 1 : (page + 1) * PAGE_SIZE]
+            ), f"fallback must fill page {page} (per-page write)"
+        # rank-0 pages (0, 2, 4) are untouched.
+        for page in (0, 2, 4):
+            assert raw[page * PAGE_SIZE] != 0 or True  # may be touched by mmap init
```
Comment (orozery) [inline tests/v1/kv_offload/cpu/test_shared_offload_region.py]: `or True` makes this assertion a no-op

File: tests/v1/kv_offload/cpu/test_shared_offload_region.py [hunk 3]
```diff
@@ -404,6 +405,94 @@ def test_file_has_correct_size(iid):
...
+    with _region(iid, num_blocks=num_blocks, num_workers=2, rank=1) as r:
+        raw = memoryview(r.mmap_obj)
+        # rank=1 occupies page 1, 3, 5 (the second column of each row).
+        for page in (1, 3, 5):
+            assert raw[page * PAGE_SIZE] == 0, (
+                f"fallback must touch rank-1 page {page} head"
+            )
+            assert all(
+                b == 0 for b in raw[page * PAGE_SIZE + 1 : (page + 1) * PAGE_SIZE]
+            ), f"fallback must fill page {page} (per-page write)"
```
Comment (orozery) [inline tests/v1/kv_offload/cpu/test_shared_offload_region.py]: Claude:

The message says "fallback must fill page" but _fallback_populate_write only writes one byte per page (at the PAGESIZE stride).
This assertion passes because ftruncate + fresh mmap gives OS-zeroed pages, not because the fallback wrote anything to bytes 1..4095.
This test doesn't verify fallback behavior — it verifies OS behavior.

**Review by orozery — COMMENTED** [review]
Claude suggestion of tests:

1. **Keep `test_madvise_success_selects_madvise_population`** — add a call-count spy on the madvise side (1 probe + N population calls).

2. **Keep `test_madvise_unexpected_oserror_propagates`** — good as-is.

3. **One EINVAL smoke test** — just verify constructor completes when madvise raises EINVAL. Drop the decorative byte assertions. One rank value is enough (iteration logic is already tested elsewhere in the file). No need for both ranked and unranked variants.

4. **Add a direct unit test for `_fallback_populate_write`** — plant non-zero sentinels, run fallback, assert sentinels survive. This is the test that catches the `= 0` bug and actually exercises the function:

```python
def test_fallback_populate_write_preserves_existing_bytes():
    size = 3 * mmap.PAGESIZE
    # create mmap, plant sentinels
    mv[0] = 0xAB
    mv[mmap.PAGESIZE] = 0xCD
    mv[2 * mmap.PAGESIZE] = 0xEF

    _fallback_populate_write(mmap_obj, 0, size)

    assert mv[0] == 0xAB       # fails with = 0, passes with |= 0
    assert mv[mmap.PAGESIZE] == 0xCD
    assert mv[2 * mmap.PAGESIZE] == 0xEF
```

**Review by orozery — APPROVED** [review]
Thanks @Alex-ai-future !
